### PR TITLE
[Data] Refactor unit tests to avoid testing directly against representation

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

Testing directly against the `repr` is brittle because 1) unit tests break if the representation format changes, even if behavior under test is otherwise correct, and 2) unit tests break if implementation details change (e.g., whether the number of rows is known prior to execution).

(This PR updates most but not all assertions against representations to test against more explicit APIs)
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_consumption.py [hunk 1]
```diff
@@ -1509,11 +1509,8 @@ def _to_pandas(ds):
 def test_column_name_type_check(ray_start_regular_shared):
     df = pd.DataFrame({"1": np.random.rand(10), "a": np.random.rand(10)})
     ds = ray.data.from_pandas(df)
-    expected_str = (
-        "MaterializedDataset(num_blocks=1, num_rows=10, "
-        "schema={1: float64, a: float64})"
-    )
-    assert str(ds) == expected_str, str(ds)
+    assert ds.count() == 10
```
Comment (omatthew98) [inline python/ray/data/tests/test_consumption.py:1512]: Nit: Maybe add an assert here for the schema too.

File: python/ray/data/tests/test_metadata_provider.py [hunk 2]
```diff
@@ -296,7 +296,6 @@ def test_default_file_metadata_provider_many_files_partitioned(
     data_path,
     endpoint_url,
     write_partitioned_df,
-    assert_base_partitioned_ds,
```
Comment (omatthew98) [inline python/ray/data/tests/test_metadata_provider.py:299]: Why was this removed?

**Review by omatthew98 — APPROVED** [review]
One nit and one question, but lgtm. This is an excellent cleanup, nice work.

File: python/ray/data/tests/test_metadata_provider.py (hunk 2 above)
Comment (bveeramani) [reply python/ray/data/tests/test_metadata_provider.py:299]: Wasn't used anywhere
